File: mobile-sdks/android/yext-analytics/src/main/java/com/yext/analytics/AnalyticsModels.kt
package com.yext.analytics

import android.os.Parcelable
import kotlinx.parcelize.Parcelize
import java.util.Date

/**
 * Coordinates representing latitude and longitude
 */
@Parcelize
data class Coordinates(
    val latitude: Double,
    val longitude: Double
) : Parcelable

/**
 * The version label of the search config
 */
enum class VersionLabel {
    PRODUCTION,
    STAGING
}

/**
 * Browser agent information
 */
@Parcelize
data class BrowserAgent(
    val browser: String? = null,
    val browserVersion: String? = null,
    val device: String? = null,
    val deviceClass: String? = null,
    val os: String? = null,
    val osVersion: String? = null,
    val userAgent: String? = null
) : Parcelable

/**
 * Chat-specific analytics fields
 */
@Parcelize
data class ChatAnalytics(
    val botId: String,
    val conversationId: String? = null,
    val responseId: String? = null
) : Parcelable

/**
 * Pages-specific analytics fields
 */
@Parcelize
data class PagesAnalytics(
    val scope: String? = null,
    val originalEventName: String? = null,
    val siteUid: Int? = null,
    val template: String? = null
) : Parcelable

/**
 * Search-specific analytics fields
 */
@Parcelize
data class SearchAnalytics(
    val experienceKey: String,
    val searchId: String? = null,
    val queryId: String? = null,
    val verticalKey: String? = null,
    val isDirectAnswer: Boolean? = null,
    val versionLabel: VersionLabel? = null,
    val versionNumber: Int? = null,
    val isGenerativeDirectAnswer: Boolean? = null
) : Parcelable

/**
 * IP address information
 */
@Parcelize
data class IPInfo(
    val address: String,
    val algorithm: String
) : Parcelable

/**
 * Monetary value information
 */
@Parcelize
data class MonetaryValue(
    val amount: Double,
    val currency: String
) : Parcelable

/**
 * Entity identifier can be either a string or integer
 */
sealed class EntityIdentifier {
    data class StringId(val value: String) : EntityIdentifier()
    data class IntId(val value: Int) : EntityIdentifier()
}

/**
 * Location information can be either coordinates or country code
 */
sealed class LocationInfo {
    data class CoordinatesLocation(val coordinates: Coordinates) : LocationInfo()
    data class CountryCodeLocation(val countryCode: String) : LocationInfo()
}

File: mobile-sdks/ios/YextAnalytics/Sources/YextAnalytics/AnalyticsModels.swift
import Foundation

/// Coordinates representing latitude and longitude
public struct Coordinates: Codable {
    public let latitude: Double
    public let longitude: Double
    
    public init(latitude: Double, longitude: Double) {
        self.latitude = latitude
        self.longitude = longitude
    }
}

/// The version label of the search config
public enum VersionLabel: String, Codable {
    case production = "PRODUCTION"
    case staging = "STAGING"
}

/// Browser agent information
public struct BrowserAgent: Codable {
    public let browser: String?
    public let browserVersion: String?
    public let device: String?
    public let deviceClass: String?
    public let os: String?
    public let osVersion: String?
    public let userAgent: String?
    
    public init(
        browser: String? = nil,
        browserVersion: String? = nil,
        device: String? = nil,
        deviceClass: String? = nil,
        os: String? = nil,
        osVersion: String? = nil,
        userAgent: String? = nil
    ) {
        self.browser = browser
        self.browserVersion = browserVersion
        self.device = device
        self.deviceClass = deviceClass
        self.os = os
        self.osVersion = osVersion
        self.userAgent = userAgent
    }
}

/// Chat-specific analytics fields
public struct ChatAnalytics: Codable {
    public let botId: String
    public let conversationId: String?
    public let responseId: String?
    
    public init(botId: String, conversationId: String? = nil, responseId: String? = nil) {
        self.botId = botId
        self.conversationId = conversationId
        self.responseId = responseId
    }
}

/// Pages-specific analytics fields
public struct PagesAnalytics: Codable {
    public let scope: String?
    public let originalEventName: String?
    public let siteUid: Int?
    public let template: String?
    
    public init(
        scope: String? = nil,
        originalEventName: String? = nil,
        siteUid: Int? = nil,
        template: String? = nil
    ) {
        self.scope = scope
        self.originalEventName = originalEventName
        self.siteUid = siteUid
        self.template = template
    }
}

/// Search-specific analytics fields
public struct SearchAnalytics: Codable {
    public let searchId: String?
    public let queryId: String?
    public let verticalKey: String?
    public let isDirectAnswer: Bool?
    public let versionLabel: VersionLabel?
    public let versionNumber: Int?
    public let experienceKey: String
    public let isGenerativeDirectAnswer: Bool?
    
    public init(
        experienceKey: String,
        searchId: String? = nil,
        queryId: String? = nil,
        verticalKey: String? = nil,
        isDirectAnswer: Bool? = nil,
        versionLabel: VersionLabel? = nil,
        versionNumber: Int? = nil,
        isGenerativeDirectAnswer: Bool? = nil
    ) {
        self.experienceKey = experienceKey
        self.searchId = searchId
        self.queryId = queryId
        self.verticalKey = verticalKey
        self.isDirectAnswer = isDirectAnswer
        self.versionLabel = versionLabel
        self.versionNumber = versionNumber
        self.isGenerativeDirectAnswer = isGenerativeDirectAnswer
    }
}

/// IP address information
public struct IPInfo: Codable {
    public let address: String
    public let algorithm: String
    
    public init(address: String, algorithm: String) {
        self.address = address
        self.algorithm = algorithm
    }
}

/// Monetary value information
public struct MonetaryValue: Codable {
    public let amount: Double
    public let currency: String
    
    public init(amount: Double, currency: String) {
        self.amount = amount
        self.currency = currency
    }
}